dallinger/experiment.py

```diff
@@ -977,117 +984,127 @@ def monitoring_statistics(self, **kw):
 
         return stats
 
-    @staticmethod
-    def count_active_complete_failed(entities):
-        n_pending = 0
-        n_completed = 0
-        n_failed = 0
-        for entity in entities:
-            if entity.failed is False:
-                is_pending = (
-                    issubclass(entity.__class__, Info) and entity.contents is None
-                )
-                if is_pending:
-                    n_pending += 1
-                else:
-                    n_completed += 1
-            else:
-                n_failed += 1
-        return n_pending, n_completed, n_failed
-
-    def network_structure(self, **kw):
-        network_ids = {i[0] for i in self.session.query(distinct(Network.id)).all()}
-        if "network_roles" in kw:
-            network_ids = {
-                i[0]
-                for i in self.session.query(distinct(Network.id)).filter(
-                    Network.role.in_(kw["network_roles"])
-                )
-            }
-        if "network_ids" in kw:
-            network_ids = network_ids.intersection(int(v) for v in kw["network_ids"])
+    def network_structure(
+        self,
+        network_roles=None,
+        network_ids=None,
+        collapsed=False,
+        transformations=False,
+    ):
+        results = {
+            "networks": self.summarize_table("network", network_roles, network_ids),
+            "nodes": self.summarize_table(
+                "node",
+                network_roles,
+                network_ids,
+                cls_filter=(lambda cls: issubclass(cls, Source)) if collapsed else None,
+            ),
+            "vectors": []
+            if collapsed
+            else self.summarize_table("vector", network_roles, network_ids),
+            "infos": []
+            if collapsed
+            else self.summarize_table("info", network_roles, network_ids),
+            "participants": [] if collapsed else self.summarize_table("participant"),
+            "trans": []
+            if collapsed or not transformations
+            else self.summarize_table("transformation", network_roles, network_ids),
+        }
 
-        network_ids = sorted(network_ids)  # default sorting is by id
+        return results
 
-        jnetworks = []
-        for network in Network.query.filter(Network.id.in_(network_ids)).all():
-            (
-                n_pending_infos,
-                n_completed_infos,
-                n_failed_infos,
-            ) = self.count_active_complete_failed(network.all_infos)
-            _, n_completed_nodes, n_failed_nodes = self.count_active_complete_failed(
-                network.all_nodes
-            )
+    def summarize_table(
+        self,
+        table: Union[Table, str],
+        network_roles: Optional[
+            List
+        ] = None,  # Restrict to objects from networks with given roles
```

Comment:
Consider writing documentation in the function and putting the comments in there. This will also make the function header more readable.

Reply:
Sure, implemented now